### ISSUE TITLE
Ledger Protocol can stall depending on the order in which states are received

### ISSUE THREAD

**Issue by lalexgap** [statechannels/go-nitro#366]
Right now when creating a new ledger proposal we look for the most recent proposal (or supported state) and use the turn number after [that](https://github.com/statechannels/go-nitro/blob/ccae3d13fc84526692fa5fcfbd1d62b146a71e77/protocols/virtualfund/virtualfund.go#L718).

When a participant goes to accept a ledger proposal they are comparing the proposal state against a supported state to make sure their funds haven't [decreased](https://github.com/statechannels/go-nitro/blob/ccae3d13fc84526692fa5fcfbd1d62b146a71e77/protocols/virtualfund/virtualfund.go#L773).

However if there are multiple proposals we haven't accepted yet this check can fail. That's because the most recent proposal could contain multiple new guarantees compared to `supported`. The virtual funding protocol is only concerned with its guarantee so the funds spent on other guarantees cause `proposedMaintainsOurFunds` to be false.

This issue is dependent on the order in which messages are received/handled. If we receive/handle the proposals in order of turn num we're ok, since `supported` is always one turn behind the proposal we're handling.

**Comment by geoknee** [comment]
This is something we missed when settling on the design for doing ledger updates for this sprint (#291). Although the leader / follower pattern prevents a race between ledger channel participants, it does not prevent a race between message handlers ( ~ multiple objective cranks) on a single participant's machine. 

This seems to underline that, as long as guarantees are added and removed from ledger channels _synchronously_, we will require a module for proposing / accepting those updates that has a view into _multiple_ virtualfund objectives, or at least has the ability to infer the intent of those objectives. This would seem to rule out the idea of making such updates entirely inside a the `virtualfund.Crank()` function. And to push us back in the direction of a queue of ledger requests which is processed separately from objective cranking.

**Comment by andrewgordstewart** [comment]
#373  includes an approach that does not require a ledger request queue, and even supports the design goal of pure objective implementations!

**Comment by kerzhner** [comment]
The leader determines the sequence of ledger updates by setting the turn number (aka version). The follower stores:
- The latest consensus turn number for a ledger channel.
- The proposed update turn number for a ledger channel. This can be stored on the objective.

When a ledger update arrives as part of the virtual fund objective, the follower has enough information to determine whether the update can be applied or is blocked on other updates.

Is the above correct? The goal of this comment is to anchor the conversation to create a common ground for discussing explicit queue vs implicit queue vs no queue.

**Comment by lalexgap** [comment]
>https://github.com/statechannels/go-nitro/pull/373 includes an approach that does not require a ledger request queue, and even supports the design goal of pure objective implementations!

Would you say #373 implicitly implements a queue by including `proposals` in the message? When we receive a message the `proposals` let us know in which order to process them.

**Comment by lalexgap** [comment]
> The leader determines the sequence of ledger updates by setting the turn number (aka version). The follower stores:
> 
> * The latest consensus turn number for a ledger channel.

Yes, this is just the last `supported` turn number for the ledger channel.

> * The proposed update turn number for a ledger channel. This can be stored on the objective.

The  **proposed update turn number** is currently not stored anywhere, instead the follower stores the expected guarantee they're looking for the in the ledger channel.

> When a ledger update arrives as part of the virtual fund objective, the follower has enough information to determine whether the update can be applied or is blocked on other updates.

Yes, the follower has enough information to decide if it can sign the proposal or needs to wait for other proposals to be accepted first.

Currently what doesn't exist is a mechanism for keeping track of objectives waiting for other proposals and cranking them once they're ready to make progress.

> Is the above correct? The goal of this comment is to anchor the conversation to create a common ground for discussing explicit queue vs implicit queue vs no queue.

**Comment by lalexgap** [comment]
>This would seem to rule out the idea of making such updates entirely inside a the virtualfund.Crank() function.

I think this also has important implications for locking and concurrency control.  Irene has two ledger channels so `virtual.Crank()` would need to acquire locks on two ledger channels before making updates. We'd need to be careful to avoid deadlock situations.

However if ledger requests are handled outside the protocol we can have a much simpler locking scheme, where we acquire a lock per ledger channel.

**Comment by andrewgordstewart** [comment]
> > #373 includes an approach that does not require a ledger request queue, and even supports the design goal of pure objective implementations!
> 
> Would you say #373 implicitly implements a queue by including `proposals` in the message? When we receive a message the `proposals` let us know in which order to process them.

I am interpreting "ledger request queue" to mean a pretty specific thing, ie. the type of protocol we have implemented in the typescript wallet. There, "ledger request queue" is a queue of internal requests that was used by the ledger protocol to decide on how to formulate the next state update -- remember that state updates would be used to fund many channels per update.

#373 does implement a queue of signed states for a ledger channel, where each state update either adds or removes a guarantee. (Removal is only hinted at, and things like top-ups also fit nicely into this pattern). The queue is stored _explicitly_ in the `ConsensusChannel` data structure. It's nothing more than an explicit way of storing state updates that record the intent of a state update.

**Comment by andrewgordstewart** [comment]
> The leader determines the sequence of ledger updates by setting the turn number (aka version). The follower stores:
> 
> * The latest consensus turn number for a ledger channel.
> * The proposed update turn number for a ledger channel. This can be stored on the objective.
> 
> When a ledger update arrives as part of the virtual fund objective, the follower has enough information to determine whether the update can be applied or is blocked on other updates.
> 
> Is the above correct? The goal of this comment is to anchor the conversation to create a common ground for discussing explicit queue vs implicit queue vs no queue.

This is more or less correct, though there's still some flexibility...

> When a ledger update arrives as part of the virtual fund objective, the follower has enough information to determine whether the update can be applied or is blocked on other updates.

For instance, if the leader always includes all unacknowledged proposals, and the follower always processes proposals in order (sorted by `turnNum`) then the follower will never be blocked on other updates. (Some updates will be redundant, and can be quickly ignored, without even checking signatures.)

**Comment by geoknee** [comment]
> Currently what doesn't exist is a mechanism for keeping track of objectives waiting for other proposals and cranking them once they're ready to make progress.

@lalexgap I think this is the nub of it. 

Let us consider an (inefficient) engine design where all objectives are cranked on a schedule (say every 1s). I think even under the current design this would solve the issue, and all of the `virtualfund` objectives would eventually become unstuck.

Assuming that messages arrive in a random order, but there are no dropped messages, the follower's engine may perform several "empty cranks" that don't do anything, until both
- the signed state with `turnNum = supportedState.turnNum + 1` arrives and is stored
- the virtualfund objective that is blocked on that state is cranked

That objective then makes progress, as well as the ledger channel's supported turn number incrementing. Now, the next objective which is blocked can make progress, and so on. 

Obviously we can cook up much more efficient ways of scheduling objective cranks (ways that reduce the number of “empty cranks”), but the "scheduled cranking" model is a pretty useful one to consider. 

I'm not sure the approach of #373 really changes the situation all that much, since it doesn't really specify how we schedule objective cranks. It may well have other advantages, but to my mind it is somewhat orthogonal to the issue at hand. Unless I am missing something.

**Comment by andrewgordstewart** [comment]
There are two important properties of the ledger protocol:

**Safety**: If my `nitro-client` signs a state update, it doesn't cause me to lose access to money.
**Liveness**: If I ask my `nitro-client` to do something, it does so in a "timely manner" (assuming reactiveness of external dependancies).

To assist the conversation, I would suggest we first agree on the required safety and liveness properties, and then evaluate solutions against those.

> > Currently what doesn't exist is a mechanism for keeping track of objectives waiting for other proposals and cranking them once they're ready to make progress.
> 
> @lalexgap I think this is the nub of it.
> 
> Let us consider an (inefficient) engine design where all objectives are cranked on a schedule (say every 1s). I think even under the current design this would solve the issue, and all of the `virtualfund` objectives would eventually become unstuck.

This would solve the **liveness** requirement under a very flexible interpretation of "timely manner". It would not, based on my understanding, solve the **safety** requirement -- I believe my `nitro-client` would currently sign a ledger update that includes some desired guarantee `G`, but also makes some other change to the outcome, like replacing another guarantee with a simple allocation.

In other words, simply fixing the liveness problem does not address the existing safety problem.

> Obviously we can cook up much more efficient ways of scheduling objective cranks (ways that reduce the number of “empty cranks”), but the "scheduled cranking" model is a pretty useful one to consider.

This would solve the **liveness** requirement under a more reasonable interpretation of "timely manner". I am arguing that there is an [obvious way](https://www.notion.so/statechannels/Concurrent-ledger-updates-e9902021110d41668085d7706f6e70dd) of scheduling cranks that addresses both the required safety and liveness property.

> I'm not sure the approach of #373 really changes the situation all that much, since it doesn't really specify how we schedule objective cranks. It may well have other advantages, but to my mind it is somewhat orthogonal to the issue at hand. Unless I am missing something.

#373 would have Irene `update` with proposals [in a specified order](https://www.notion.so/statechannels/Concurrent-ledger-updates-e9902021110d41668085d7706f6e70dd#e3fac17e668b45a8b2d1a8769f4fbf7a).

**Comment by geoknee** [comment]
I agree that Safety and Liveness are both important. I wasn't under the impression there was a problem with **safety**, at the moment? I think the `virtualfund.Crank()` function performs enough checks to avoid me losing money, given some assumptions about locking of ledger channels. Can you expand a bit more on what you see as the problem?

**Comment by andrewgordstewart** [comment]
> I agree that Safety and Liveness are both important. I wasn't under the impression there was a problem with **safety**, at the moment? I think the `virtualfund.Crank()` function performs enough checks to avoid me losing money, given some assumptions about locking of ledger channels. Can you expand a bit more on what you see as the problem?

To be honest, I find it very difficult to fully understand what is happening within `virtualfund.crank`, but this is my general understanding:
1. If `ledger.Proposed()` returns `true`, I would call `acceptLedgerUpdate` [here](https://github.com/statechannels/go-nitro/blob/36203635b6985d4c3db28cd926754d3b80e347b0/protocols/virtualfund/virtualfund.go#L644). (It's not obvious to follow what `ledger.Proposed()` does, but we can probably assume it's correct)
2. When accepting a ledger update, I [get the latest proposed state](https://github.com/statechannels/go-nitro/blob/36203635b6985d4c3db28cd926754d3b80e347b0/protocols/virtualfund/virtualfund.go#L585)
3. I check that the proposed update ["maintains our funds"](https://github.com/statechannels/go-nitro/blob/36203635b6985d4c3db28cd926754d3b80e347b0/protocols/virtualfund/virtualfund.go#L609)
4. I check that the proposed update ["affords the guarantee"](https://github.com/statechannels/go-nitro/blob/36203635b6985d4c3db28cd926754d3b80e347b0/protocols/virtualfund/virtualfund.go#L611) I wanted included
5. If yes, I sign the proposed update (and my signature is sent out to the world)

The problem appears to be in (4): what happens if the update affords the guarantee, but replaces another guarantee with a `{destination: eve, amount: 100}` allocation item? The code currently seems to check that _the changes I want to make were made_, but doesn't seem to check that _changes that I don't want to happen didn't happen_.


There's a few things I learned looking into this:

## A. Our tests can be hard to read

I considered checking my belief by writing a failing test, but [this file](https://github.com/statechannels/go-nitro/blob/36203635b6985d4c3db28cd926754d3b80e347b0/protocols/virtualfund/virtualfund-single-hop_test.go) is pretty intimidating!

## B. Our code could be cleaner

- Logic is spread out somewhat: eg. `outcome.go` contains a lot of functionality that seems like it only needs to be used by `virtualfund.go`
- inconvenient data structures are used (see [here](https://github.com/statechannels/go-nitro/pull/382#discussion_r833355829))

## C. abstract code like `Channel.addSignStates`, `Channel.signedStateForTurnNumber` can make simple computations more difficult to implement.

Compare [`Ledger.Proposed()`](https://github.com/statechannels/go-nitro/blob/36203635b6985d4c3db28cd926754d3b80e347b0/channel/ledger.go#L46-L62) with
```
func (c ConsensusChannel) Proposed() Vars {
	result := c.current.Vars

	for _, p := range c.proposalQueue {
		result, _ = result.Add(p.Proposal.(Add))
	}

	return result
}
```

## D. abstract code like `Channel.addSignStates`, `Channel.signedStateForTurnNumber` can make it harder to implement correct functionality

Receiving a generic state in `virtualFund.update` + calling `addSignStates` implies my client may put pretty arbitrary data in the store. This means that when I call `virtualFund.crank`, I need to do a lot of validation on the data I have to make sure my channel is in a "safe state" before proceeding. (I'm not sure we always do that validation)

A related example is the problem I outlined in 1-5:
- currently, we use abstract `addSignedState` code to validate and store arbitrary signed states that are sent across the wire
- some of those state (updates) have special meaning. the special meaning is "lost in the mix" by putting them into a bag of states
- the net result is that it is non-trivial to validate that funds are invariant. We would have to assert that states are sequential, sort by turnNum, use some [special knowledge](https://github.com/statechannels/go-nitro/blob/36203635b6985d4c3db28cd926754d3b80e347b0/protocols/virtualfund/virtualfund.go#L596-L611) to unpack a generic `outcome.Exit`, and be very careful when analyzing the diff to check that both a) the _right_ change is made to the `outcome.Exit` and b) _no other_ change is made to the `outcome.Exit`

Compare this with the following algorithm:
**Leader:** 
1. Construct a guarantee `g` that I want to include
2. Compute `nextState = currentState.divertFundsTo(g)`
3. Sign `nextState` and send `(g, signature)` across the wire

**Follower:** 
1. Receive `(g, signature)`. If `g` looks funny, abort.
2. Compute `nextState = currentState.divertFundsTo(g)`
3. Recover the signer from `(nextState, g)` and assert that the signer was `L.leader`

This algorithm is straightforward and less error-prone: `nextState` is known to the follower to maintain the follower's total funds, because it was computed from a `currentState` that they trust (because they signed it, for instance) + a change that they agree with (because their `virtualfund` objective says it's ok). The follower rejects any signature that doesn't match with the precise change that the `virtualfund` objective wants to see.

**Comment by geoknee** [comment]
I see. I think you're right that the current code is not safe against some malicious behaviour by other parties. 

I also agree that we can make our code much better (less bug prone, more terse, etc) by using more ergonomic data structures. I like your proposed algorithm to solve the problem of countersigning updates that may include changes I do not wish to authorise.
